lib/Doctrine/DBAL/Cache/ResultCacheStatement.php

```diff
@@ -179,18 +180,26 @@ public function fetch($fetchMode = null, $cursorOrientation = PDO::FETCH_ORI_NEX
      */
     public function fetchAll($fetchMode = null, $fetchArgument = null, $ctorArgs = null)
     {
-        $data = $this->statement->fetchAll($fetchMode, $fetchArgument, $ctorArgs);
-
-        if ($fetchMode === FetchMode::COLUMN) {
-            foreach ($data as $key => $value) {
-                $data[$key] = [$value];
-            }
-        }
+        $data = $this->statement->fetchAll(FetchMode::ASSOCIATIVE, $fetchArgument, $ctorArgs);
 
         $this->data    = $data;
         $this->emptied = true;
 
-        return $this->data;
+        if ($fetchMode === FetchMode::NUMERIC) {
+            foreach ($data as $i => $row) {
+                $data[$i] = array_values($row);
+            }
+        } elseif ($fetchMode === FetchMode::MIXED) {
+            foreach ($data as $i => $row) {
+                $data[$i] = array_merge($row, array_values($row));
+            }
+        } elseif ($fetchMode === FetchMode::COLUMN) {
+            foreach ($data as $i => $row) {
+                $data[$i] = reset($row);
+            }
+        }
+
+        return $data;
```

Comment:
How about this:

```suggestion
        return array_map(function(array $row) use ($fetchMode)  {
            switch ($fetchMode)
            {
                case FetchMode::NUMERIC:
                    return array_values($row);

                case FetchMode::MIXED:
                    return array_merge($row, array_values($row));
                
                case FetchMode::COLUMN:
                    return reset($row);
                
                default:   
                    return $row;
            }
        }, $data);
```
?

Reply:
It will require a switch on each row instead of an if/else once.